symengine/polys/uintpoly.cpp

```diff
@@ -1,7 +1,311 @@
+                         unsigned int l) const
+{
+    std::vector<UIntDict> res;
+    size_t r = f_list.size();
+    auto lc = get_lc();
+
+    if (r == 1) {
+        integer_class g, r, s, pl;
+        mp_pow_ui(pl, p, l);
+        mp_gcdext(g, r, s, lc, pl);
+        UIntDict F((*this) * r);
+        F.itrunc(pl);
+        res.push_back(F);
+        return res;
+    }
+
+    std::vector<UIntDict> sub1, sub2;
+    integer_class m(p);
+    size_t k(r / 2);
+    auto d = std::ceil(std::log2(l)); // TODO
+    auto g = GaloisFieldDict::from_vec({lc}, p);
+
+    for (unsigned i = 0; i < k; ++i) {
+        g *= GaloisFieldDict(f_list[i].dict_, p);
+        sub1.push_back(f_list[i]);
+    }
+
+    sub2.push_back(f_list[k]);
+    auto h = GaloisFieldDict(f_list[k].dict_, p);
+    for (unsigned i = k + 1; i < f_list.size(); ++i) {
+        h *= GaloisFieldDict(f_list[i].dict_, p);
+        sub2.push_back(f_list[i]);
+    }
+
+    GaloisFieldDict temp2, s, t;
+    GaloisFieldDict::gf_gcdex(g, h, outArg(s), outArg(t), outArg(temp2));
+
+    UIntDict dg = UIntDict::from_vec(g.dict_);
+    dg.itrunc(p);
+    UIntDict dh = UIntDict::from_vec(h.dict_);
+    dh.itrunc(p);
+    UIntDict ds = UIntDict::from_vec(s.dict_);
+    ds.itrunc(p);
+    UIntDict dt = UIntDict::from_vec(t.dict_);
+    dt.itrunc(p);
+
+    for (unsigned int i = 1; i <= d; ++i) {
+        UIntDict::zz_hensel_step(m, (*this), dg, dh, ds, dt, outArg(dg),
+                                 outArg(dh), outArg(ds), outArg(dt));
+        mp_pow_ui(m, m, 2);
+    }
+
+    res = dg.zz_hensel_lift(p, sub1, l);
+    auto temp = dh.zz_hensel_lift(p, sub2, l);
+    res.insert(res.end(), temp.begin(), temp.end());
+    return res;
+}
+
+void UIntDict::zz_divide(const UIntDict &a, const UIntDict &b,
+                         const Ptr<UIntDict> &quo, const Ptr<UIntDict> &rem)
+{
+    auto b_poly = b;
+    if (a.size() == 0)
+        throw std::runtime_error("Zero Division Error");
+    map_uint_mpz res;
+    UIntDict tmp;
+    integer_class q, r;
+    unsigned int a_deg, b_deg;
+    while (b_poly.degree() >= a.degree()) {
+        a_deg = a.degree();
+        b_deg = b_poly.degree();
+        q = b_poly.get_lc() / a.get_lc();
+        res[b_deg - a_deg] = q;
+        tmp = UIntDict(map_uint_mpz{{b_deg - a_deg, q}});
+        b_poly -= (a * tmp);
+    }
+    *quo = UIntDict(res);
+    *rem = b;
+    *rem -= (a * (*quo));
+}
+
+std::set<RCP<const UIntPoly>, RCPBasicKeyLess> UIntPoly::zz_zassenhaus() const
+{
+    std::set<RCP<const UIntPoly>, RCPBasicKeyLess> factors;
+    unsigned int n = get_degree();
+    UIntDict f(this->poly_);
+    if (n == 1) {
+        factors.insert(rcp_from_this_cast<const UIntPoly>());
+        return factors;
+    }
+    integer_class fc(this->poly_.dict_.begin()->second);
+    auto A = mp_get_ui(this->poly_.max_abs_coef());
+    auto b = get_lc();
+    auto B
+        = std::abs(int(std::sqrt(n + 1)) * std::pow(2, n) * A * mp_get_si(b));
```

Comment:
When we do hensel lifting.
[Here](
https://github.com/nishnik/symengine/blob/hensel/symengine/polys/uintpoly.cpp#L106),
we convert the polynomial from `Z` to a polynomial in field `p**l`.
Then we do hensel lifting on all the factors of field `p`, and then [here](
https://github.com/nishnik/symengine/blob/hensel/symengine/fields.cpp#L1043)
we lift all the factors to field of `p**l`.
Though I see that, in the previous implementation, where we used polynomial
division in integer domain, this problem is solved. As we are using
`UIntPoly` representation, which is sparse there is no issue of space
(Though I am unsure of performance).
And [here](
https://github.com/nishnik/symengine/blob/hensel/symengine/polys/uintpoly.cpp#L93)
we are implementing `zassenhaus` algorithm in field `p`, so a smaller
vector is required.
I personally think that we should move back to previous
implementation(SymPy does that, though its division is also in dense
polynomials).
I will look into the performance matter tonight and update you.
cc @certik

On Thu, Aug 18, 2016 at 11:16 PM, Isuru Fernando <notifications@github.com>
wrote:

> In symengine/polys/uintpoly.cpp
> <https://github.com/symengine/symengine/pull/1066#discussion_r75354742>:
>
> > +}
> > +
> > +std::set<RCP<const UIntPoly>, RCPBasicKeyLess> UIntPoly::zz_zassenhaus() const
> > +{
> > +    std::set<RCP<const UIntPoly>, RCPBasicKeyLess> factors;
> > +    unsigned int n = get_degree();
> > +    UIntDict f(this->poly_);
> > +    if (n == 1) {
> > +        factors.insert(rcp_from_this_cast<const UIntPoly>());
> > +        return factors;
> > +    }
> > +    integer_class fc(this->poly_.dict_.begin()->second);
> > +    auto A = mp_get_ui(this->poly_.max_abs_coef());
> > +    auto b = get_lc();
> > +    auto B
> > +        = std::abs(int(std::sqrt(n + 1)) * std::pow(2, n) * A * mp_get_si(b));
>
> Where is the dictionary of size pl made?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/symengine/symengine/pull/1066/files/e76c7a23a010d458384a14abb2973342023e1772#r75354742>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI0uix0ExEpYj08AUdLIu1Y27wOEh5c6ks5qhJqTgaJpZM4Jl-E3>
> .
>